File: N皇后II.cpp
class Solution {
public:
    int totalNQueens(int n) {
        int ans = 0;
        vector<bool> col(n * 2 + 1, false), dig(n * 2 + 1, false), back_dig(n * 2 + 1, false);
        vector<string> g(n);

        function<void(int)> dfs = [&](int deep) -> void {
            if (deep == n) {
                ans++;
                return;
            }
            for (int i = 0; i < n; ++i) {
                if(!(col[i] || dig[n + deep - i] || back_dig[deep + i])) {
                    col[i] = dig[n + deep - i] = back_dig[deep + i] = true;
                    dfs(deep + 1);
                    col[i] = dig[n + deep - i] = back_dig[deep + i] = false;
                }
            }
        };

        dfs(0);
        return ans;
    }
};

File: N皇后II.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        ans = 0
        col, dig, back_dig = [0 for _ in range(n * 2 + 1)], [0 for _ in range(n * 2 + 1)], [0 for _ in range(n * 2 + 1)]
        
        def dfs(deep):
            if deep == n:
                nonlocal ans
                ans += 1
                return
            for i in range(n):
                if (not (col[i] or dig[n + deep - i] or back_dig[deep + i])):
                    col[i] = dig[n + deep - i] = back_dig[deep + i] = 1
                    dfs(deep + 1)
                    col[i] = dig[n + deep - i] = back_dig[deep + i] = 0
        
        dfs(0)
        return ans
